### DISCUSSION TITLE
Support for RFC 9421 - HTTP Message Signatures

### DISCUSSION THREAD

**Discussion by forty** [curl/curl#13376] [Ideas]
Hello,

Is there any interest to add support for [RFC 9421 - HTTP Message Signatures](https://datatracker.ietf.org/doc/html/rfc9421) to curl ? I saw that curl already support AWS Signature v4 (with --aws-sigv4), and I feel this RFC is a generalization / standardisation of such signature (I notice participation of amazon to the RFC, which I take as a sign it could be a successor to SIGv4). Is there any past discussion on this topic?

The RFC seems super wide and flexible, so I think an interesting first step on this would be to define a subset that realistically usable from the CLI curl, and clearly define that subset so server people could implement the subset RFC that's usable from curl (or maybe define "profiles" which define what is signed, or something like this, to avoid having to list all the headers, etc)

(we have our own old home made signature system, which means we currently cannot use curl for our API, but with this one being standardize, I feel this is an opportunity to open this discussion, so people who sign their requests can finally do curl --rfc9421 -u "user:key" https://... :D )

**Comment by bagder** [comment]
Hello!

This functionality certainly seems like something that would be suitable for curl to support. I assume that as soon as there is server support and some clients start to use HTTP message signatures, people will come to realize and think of use cases for when curl should as well. After all, having a clear use case is often a good driver for how to implement features.

**Reply by pointlessone** [reply]
ActivityPub uses message signatures extensively.

**Reply by jricher** [reply]
@pointlessone ActivityPub uses a pre-RFC predecessor to RFC9421, which comes in various mutually-incompatible versions and is not compatible with the final spec. I'm not active in the ActivityPub space but there seems to be a conversation happening about that here: https://socialhub.activitypub.rocks/t/towards-an-http-signature-fep/3896

**Reply by gobengo** [reply]
@jricher 

> ActivityPub uses a pre-RFC predecessor to RFC9421,

That's not quite true. I think you'll find that https://www.w3.org/TR/activitypub/ does not mention any pre-RFC predecessor to RFC9421.

The TR just refers to 'http signatures', any other links were removed before 2018 so that that reference could refer to the RFC when done.

I think it's just as accurate to say "ActivityPub uses RFC9421" as to say it uses a pre-rfc predecessor.

Specific software projects out there, regardless of what ActivityPub says, may use pre-rfc predcecessors, but it's not because the ActivityPub protocol recommends or requires it.

**Reply by kiler129** [reply]
I think this section summarizes it nicely: https://swicg.github.io/activitypub-http-signature/#how-to-upgrade-supported-versions. I think saying *"pre-rfc RFC 9421"* loosely refers to [a draft by Cavage that expired in 2018](https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures), but has been in active use for years for signatures handling.

Researching this topic right now in 11/2025, I stumbled upon a [good note by OAuth.net](https://oauth.net/http-signatures/) as *"how we got here"* regarding RFC 9421:

> *This draft is an evolution of the earlier individual draft "[Signing HTTP Messages](https://tools.ietf.org/id/draft-cavage-http-signatures-08.html)" by Cavage. The Cavage draft was never adopted by a working group and expired in 2018. For a brief period of time, some of this work had been picked up by the "[Digital Verification Community Group](https://github.com/w3c-dvcg)" and later moved into the "[Credentials Community Group](https://github.com/w3c-ccg)" at the W3C. The [CCG version](https://github.com/w3c-ccg/http-signatures/blob/gh-pages/index.xml) of the draft was redirected to the IETF draft in the [HTTPBIS working group](https://datatracker.ietf.org/doc/draft-ietf-httpbis-message-signatures/) where development of the draft continued until it was published as an RFC.*

---

IMO, since curl never implemented signatures verification, it shouldn't bother with implementing any of the legacy/draft signatures. However, documentation for signature verification failure code emitted should probably provide some of the historical context to help users debug why their signature verification doesn't work. The situation reminds me a bit of the 5 incompatible-but-used versions of WebSocket back in the day.

**Reply by aaronpk** [reply]
Thank you, it took me quite a while to trace all the history in order to capture it in that paragraph on oauth.net!

ActivityPub doesn't reference any version of HTTP Signatures directly. Mastodon and most ActivityPub implementations on the other hand do absolutely work with only a pre-RFC9421 version of HTTP Signatures.

Regardless, curl should definitely only implement the RFC version of HTTP Signatures, as that has gone through a much more rigorous review process and adheres to HTTP standards and conventions.

**Comment by jricher** [comment]
Hi, editor for RFC9421 here! Support for this spec in curl is a great idea, especially as there is precedent with SigV4. On the server side, we're more likely to see support up at the application/framework layer as opposed to, say, apache and nginx.

For developers, I also want to point out the resource https://httpsig.org/ for both testing and library support. The spec itself also has a lot of test vectors in it that have been confirmed by multiple independent implementations.

**Comment by jeremiahlee** [comment]
[RFC 9635](https://www.ietf.org/rfc/rfc9635.html), the Grant Negotiation and Authorization Protocol (GNAP), [uses](https://www.ietf.org/rfc/rfc9635.html#name-http-message-signatures) HTTP Message Signatures.

An example of GNAP in use is the [Open Payments API specification](https://openpayments.dev/introduction/overview/#grant-negotiation-and-authorization). Its public test environment: [Grant Request endpoint](https://openpayments.dev/apis/auth-server/operations/post-request/): https://auth.interledger-test.dev/

**Comment by rnett** [comment]
We are now shipping HTTP message signature based authentication in our application (as an alternative to mTLS that works with middleware), so having curl support it would be very helpful.  We use a hard-coded list of components in the signature, including some custom headers, with configurable key and cert files.  Our ideal use-case for cURL would be for customers who use HTTP message signature based auth to be able to use it with our API, with some minimal configuration (e.g. a config file template provided by us).

There is an [HTTPie plugin](https://pypi.org/project/httpie-http-message-signatures/0.0.2/) that adds support for message signatures using a config file, which would work well for us.

**Comment by bagder** [comment]
See https://github.com/curl/curl/pull/21239

**Reply by shalk** [reply]
The pr is closed . any updates ?

**Reply by bagder** [reply]
it has been merged! [details on why it is "closed" and not "merged"](https://daniel.haxx.se/blog/2024/06/11/why-curl-closes-prs-on-github/)

The blog post about it: https://daniel.haxx.se/blog/2026/07/27/http-message-signatures-with-curl/

**Reply by shalk** [reply]
That is great ~

**Comment by dfandrich** [comment]
https://daniel.haxx.se/blog/2024/06/11/why-curl-closes-prs-on-github/
